### PR TITLE
RPC: Migrate estimatesmartfee return value from "feerate" (BTC/kvB) to "fee_rate" (sat/vB)

### PR DESCRIPTION
**Pull request by luke-jr**
Not sure if it makes sense to do this or not...

### CODE REVIEW

**Comment by kristapsk** [comment]
I don't think deprecating `feerate` here so fast is a good idea, as `estimatesmartfee` is widely used RPC by various software. Maybe just add `fee_rate` here and deprecated `feerate` at some point later in the future?

File: src/policy/feerate.h [hunk 1]
```diff
@@ -65,7 +66,10 @@ class CFeeRate
     friend bool operator>=(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK >= b.nSatoshisPerK; }
     friend bool operator!=(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK != b.nSatoshisPerK; }
     CFeeRate& operator+=(const CFeeRate& a) { nSatoshisPerK += a.nSatoshisPerK; return *this; }
+    /** Returns the fee rate in BTC/kvB or sat/vB, with units, as a string */
     std::string ToString(const FeeEstimateMode& fee_estimate_mode = FeeEstimateMode::BTC_KVB) const;
+    /** Returns the fee rate in sat/vB, without units, as a string */
+    std::string SatsToString() const { return strprintf("%d.%03d", nSatoshisPerK / 1000, nSatoshisPerK % 1000); }
```
Comment (jonatack) [inline src/policy/feerate.h:72]: LMK if you have a better idea how to name this helper :D

File: src/rpc/mining.cpp [hunk 2]
```diff
@@ -1041,7 +1041,8 @@ static RPCHelpMan estimatesmartfee()
                 RPCResult{
                     RPCResult::Type::OBJ, "", "",
                     {
-                        {RPCResult::Type::NUM, "feerate", /* optional */ true, "estimate fee rate in " + CURRENCY_UNIT + "/kB (only present if no errors were encountered)"},
+                        {RPCResult::Type::NUM, "feerate", /* optional */ true, "estimate fee rate in " + CURRENCY_UNIT + "/kB (only present if no errors were encountered; DEPRECATED, also returned only if the config option -deprecatedrpc=feerate is passed)"},
+                        {RPCResult::Type::NUM, "fee_rate", /* optional */ true, "estimate fee rate in " + CURRENCY_ATOM + "/vB (only present if no errors were encountered)"},
```
Comment (jonatack) [inline src/rpc/mining.cpp:1045]: Would need to also update the help further up, e.g.
```diff
+++ b/src/rpc/mining.cpp
@@ -1022,7 +1022,7 @@ static RPCHelpMan submitheader()
 static RPCHelpMan estimatesmartfee()
 {
     return RPCHelpMan{"estimatesmartfee",
-                "\nEstimates the approximate fee per kilobyte needed for a transaction to begin\n"
+                "\nEstimates the approximate fee rate in satoshis per vbyte (sat/vB) needed for a transaction to begin\n"
                 "confirmation within conf_target blocks if possible and return the number of blocks\n"
                 "for which the estimate is valid. Uses virtual transaction size as defined\n"
                 "in BIP 141 (witness data is discounted).\n",
@@ -1031,7 +1031,7 @@ static RPCHelpMan estimatesmartfee()
                     {"estimate_mode", RPCArg::Type::STR, /* default */ "CONSERVATIVE", "The fee estimate mode.\n"
             "                   Whether to return a more conservative estimate which also satisfies\n"
             "                   a longer history. A conservative estimate potentially returns a\n"
-            "                   higher feerate and is more likely to be sufficient for the desired\n"
+            "                   higher fee rate and is more likely to be sufficient for the desired\n"
             "                   target, but is not as responsive to short term drops in the\n"
             "                   prevailing fee market.  Must be one of:\n"
```

**Review by jonatack — COMMENTED** [review]
Concept ACK, ISTM this is the direction we need to go in for #19543. The last two commits need updated tests/deprecation test.

**Comment by jonatack** [comment]
> I don't think deprecating `feerate` here so fast is a good idea, as `estimatesmartfee` is widely used RPC by various software. Maybe just add `fee_rate` here and deprecate `feerate` at some point later in the future?

I agree, but here are two reasons why deprecating immediately may (a) be acceptable to reviewers and (b) a good idea:

(a) I made a similar argument a couple months ago against immediately deprecating the getpeerinfo "addnode" field at the same time as a new field was added to replace it ("addnode": true/false -> "connection_type": "manual"/other string values); reviewers disagreed and preferred to deprecate immediately and have it done with

(b) More importantly IMO, having both `feerate` and `fee_rate` with different units is not only a bit confusing but also a potential footgun, albeit a hopefully benign one with the 1E5 downward difference between the two units.

**Comment by MarcoFalke** [comment]
What is the benefit of forcing every user to change their scripts when both units can be supported forever at no additional cost?

**Comment by jonatack** [comment]
> What is the benefit of forcing every user to change their scripts when both units can be supported forever at no additional cost?

Maybe this is answered by my comment just above.

An alternative would be to rename them and add the units in the names, but I think that's sub-optimal if we want to simplify and standardize the naming and units.

I took another approach with #20391: as settxfee, like estimatesmartfee, is arguably slightly misnamed (they are really about a fee rate, not a fee), I proposed a `setfeerate` RPC in sat/vB (that also aims to provide an improved user experience to encourage its adoption).

**Comment by sipa** [comment]
If the goal is user confusion rather than maintenance burden, we could just make the new fields preferred over the old ones, and e.g. stop listing the old ones in help output.

**Comment by kristapsk** [comment]
> What is the benefit of forcing every user to change their scripts when both units can be supported forever at no additional cost?

That's the approach c-lightning is taking with various calls.

This isn't affecting just some scripts, I think this may affect most of the Bitcoin software in existence that relies on Bitcoin Core. And I believe [not breaking userspace](https://github.com/bitcoin/bitcoin/issues/14363) is very important.

**Comment by jonatack** [comment]
> What is the benefit of forcing every user to change their scripts when both units can be supported forever at no additional cost?

Remember, the scripts don't have to change if the deprecation flag is passed on restart after updating to 0.21. They have to change only if the code is actually removed later on. The deprecation just hides the fields and help somewhat similarly to @sipa's suggestion.

**Comment by kristapsk** [comment]
> scripts don't have to change if the deprecation flag is passed on restart after updating to 0.21.

That might only help if Bitcoin Core is bundled into other software, user uses his own scripts or is Bitcoin dev in general.

Ordinary user may use Bitcoin Core, c-lightning, JoinMarket and other software relying on Bitcoin Core on his computer, updating everything or parts with OS package manager. They might miss such breaking API changes if they happen so fast, you can't expect everybody to read all the release notes. When more time passes, more software might by updated to use new API and it would be safer to start deprecating old behaviour then.

**Comment by jonatack** [comment]
Ok, to not spend more time on this, please confirm if you think this would be safe and clear or if we should stop normalizing to sat/vB.

```json
{
  "feerate": 0.00078464,
  "fee_rate": 78.464,
  "blocks": 6
}
```

**Comment by kristapsk** [comment]
@jonatack , yes that looks ok to me, should not break any properly written software.

**Comment by jonatack** [comment]
Thanks. Hmmm.

> I took another approach with #20391: as settxfee, like estimatesmartfee, is arguably slightly misnamed (they are really about a fee rate, not a fee), I proposed a `setfeerate` RPC in sat/vB (that also aims to provide an improved user experience to encourage its adoption).

I might propose a separate `estimatesmartfeerate` RPC in sat/vB.

**Comment by jonatack** [comment]
Thinking about this more, I propose we:

- add sat/vB RPCs `setfeerate` (done in #20391), `estimatefeerate`, and maybe `estimaterawfeerate`
- deprecate `feeRate` (done in #20483)
- make `settxfee` and `estimatesmartfee` hidden RPCs to avoid confusion for new users while allowing current ones to keep using them

This seems the best way to: 

- normalize and transition to sat/vB per user demand
- protect users by not having `feeRate/feerate` in BTC/kvB and `fee_rate` in sat/vB in the same RPCs to avoid user confusion and error
- minimize the amount of breaking API changes

**Comment by kristapsk** [comment]
@jonatack  , sounds like a good plan to me!

**Comment by jonatack** [comment]
@kristapsk great! I'll do it today.

**Comment by luke-jr** [comment]
I like @jonatack 's plan/approach better, closing.

**Pull request closed without merging**